node-repository/src/main/java/com/yahoo/vespa/hosted/provision/node/filter/NodeOsVersionFilter.java

```diff
@@ -0,0 +1,35 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.provision.node.filter;
+
+import com.yahoo.component.Version;
+import com.yahoo.vespa.hosted.provision.Node;
+
+import java.util.Objects;
+
+/**
+ * Filter nodes by their OS version.
+ *
+ * @author mpolden
+ */
+public class NodeOsVersionFilter extends NodeFilter {
+
+    private final Version version;
+
+    private NodeOsVersionFilter(Version version, NodeFilter next) {
+        super(next);
+        this.version = Objects.requireNonNull(version, "version cannot be null");
+    }
+
+    @Override
+    public boolean matches(Node node) {
+        if (!version.isEmpty() && !node.status().osVersion().filter(v -> v.equals(version)).isPresent()) {
```

Comment:
I found this statement a bit hard to read, would this be equivalent?
``
        Optional<Version> nodeVersion = node.status().osVersion();
        if (this.version.isEmpty() || !nodeVersion.isPresent()) return nextMatches(node);
        return this.version.equals(nodeVersion.get());
``

Reply:
Thanks, that's better. Will change it.

Reply:
Actually, that's not the same thing due to the behavior of `NodeFilter#nextMatches`.

Your suggestion would be fine if `NodeFilter#nextMatches` method didn't return `true` for the last filter in the chain. I think the intention is that filters should match a given node if no filters in the chain actively reject it (a bit strange, but other code depends on this behavior).